lib/rule_engine/ast.py

```diff
@@ -377,14 +381,28 @@ def __op_arithmetic(self, op, thing):
 		_assert_is_numeric(right_value)
 		return op(left_value, right_value)
 
-	_op_add  = functools.partialmethod(__op_arithmetic, operator.add)
 	_op_sub  = functools.partialmethod(__op_arithmetic, operator.sub)
 	_op_fdiv = functools.partialmethod(__op_arithmetic, operator.floordiv)
 	_op_tdiv = functools.partialmethod(__op_arithmetic, operator.truediv)
 	_op_mod  = functools.partialmethod(__op_arithmetic, operator.mod)
 	_op_mul  = functools.partialmethod(__op_arithmetic, operator.mul)
 	_op_pow  = functools.partialmethod(__op_arithmetic, operator.pow)
 
+class AddExpression(LeftOperatorRightExpressionBase):
+	"""A class for representing arithmetic expressions from the grammar text such as addition and subtraction."""
+	compatible_types = (DataType.FLOAT,DataType.STRING)
+	result_type =  DataType.STRING
```

Comment:
i didn't really dig into how you implemented `result_type` here, but this should actually probably be `(DataType.STRING, DataType.FLOAT)`.

Reply:
It's used for the type hinting so exceptions can be raised at parse time before evaluation. If it's supposed to be multiple data types, you should start it out as `UNDEFINED` and then set it in the `__init__` method. You can see an example of this pattern in the `BitwiseExpression` class. That's probably pretty similar to what's going on here. Some bitwise operators can be applied when both the left and right values either either FLOAT or SET.